Setup TLS with CA Cert for vsphere cloud provider

mariantalla opened the pull request:
- Extend config to take a path to a CA Certificate
- Use the CA Cert when establishing a connection with the SOAP client

Testing
We provide certs and keys for tests as fixtures, `vclib/fixtures`.
Those were created (and can be regenerated) using `vclib/fixtures/createCerts.sh`.

At the moment it's possible to configure a CA path and at the same time allow insecure
communication between vsphere cloud provider and vcenter. This may
change in the future; we might opt for overwriting the insecure
communication if a CA is configured / log and transparently pass the
arguments to the vcenter command / other. To be discussed.

At the moment the CA is a global level configuration. In other
words, all vcenter servers need to use certificates signed by the same
CA. There might be use cases for different CA per vcenter server; to be
discussed.

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This PR adds the option of configuring a trusted CA for the communication between the vsphere cloud provider and the vcenter control plane.

**Which issue(s) this PR fixes**:
Fixes #64222

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
- Can configure the vsphere cloud provider with a trusted Root-CA
```

Commit:
Setup TLS with CA Cert

- Extend config to take a path to a CA Certificate
- Use the CA Cert when establishing a connection with the SOAP client

Testing
We provide certs and keys for tests as fixtures, `vclib/fixtures`.
Those were created (and can be regenerated) using `vclib/fixtures/createCerts.sh`.

At the moment it's possible to configure a CA path and at the same time allow insecure
communication between vsphere cloud provider and vcenter. This may
change in the future; we might opt for overwriting the insecure
communication if a CA is configured / log and transparently pass the
arguments to the vcenter command / other. To be discussed.

At the moment the CA is a global level configuration. In other
words, all vcenter servers need to use certificates signed by the same
CA. There might be use cases for different CA per vcenter server; to be
discussed.

wgliang commented on pkg/cloudprovider/providers/vsphere/vclib/connection_test.go:
```diff
@@ -0,0 +1,122 @@
+package vclib_test
```
LICENSE and

package vclib

wgliang replied:
TableDrivenTests and more test cases will be better.

wgliang commented on pkg/cloudprovider/providers/vsphere/vsphere_test.go:
```diff
@@ -89,8 +91,12 @@ func configFromEnv() (cfg VSphereConfig, ok bool) {
+	return configFromSimWithTLS(new(tls.Config), true)
+}
+
+// configFromSim starts a vcsim instance and returns config for use against the vcsim instance.
```
comment is wrong.

Commit:
Make bazel happy

./hack/update-bazel.sh

Commit:
Add LICENSE header

Commit:
Improve godocs and testcase naming

hoegaarden commented:
/ok-to-test

mariantalla commented:
@wgliang Thanks for the feedback. For some reason your comments disappeared when we pushed new commits.

#### Some points you brought up that should be hopefully sorted now:
- `LICENCE` for test files added
- comments that document functions `configFromSim` and `configFromSimWithTLS` are now accurate

#### Remaining points that you brought up:
- More test cases: what would you like to see covered which is not already?
- TableDrivenTests: we think that it will introduce more overhead than help at the moment, but will certainly consider it once we have more test cases (we still have work to do on this PR).
- `vclib_test` vs `vclib` package for tests: We generally tend to lean towards blackbox testing, so we force ourselves to not touch internals of packages. Thus the package `vclib_test`. What were your thoughts on putting the tests in the same package as the source?

Commit:
Add LICENCE header to createCerts.sh

Also remove comments that are not useful anymore.

wgliang commented:
@mariantalla 
I'm very sorry, because I didn't see WIP at first, you can change it to the more obvious [WIP].

mariantalla commented:
@wgliang Thanks, changed the title! Please keep commenting even though it's still in progress. Would love to hear your thoughts on test cases and packages in particular.

wgliang commented:
@mariantalla 
Of course, thank you very much and willingly.

Commit:
Fix spelling

Commit:
Resolve paths of test fixtures at runtime

This will help with bazel tests, which seem to use a different working
directory from local test runs.

Commit:
Add godocs for fixtures

... and rename `InvalidCaCertPath` to `InvalidCertPath`.

Commit:
Introduce thumbprints per vcenter

Commit:
Pass through CA cert file to the connection when multiple vcenters are configured

Commit:
Setup test for verifying by checking certificate fingerprints

Commit:
Use soap clients method to load root CAs

Commit:
Check certificate thumbprint when configured

hoegaarden committed:
Create new certs & cleanup cert generation

The test certificates are now valid for 200 years.
The generation script has been cleaned up a bit.

hoegaarden commented:
/test pull-kubernetes-node-e2e

hoegaarden commented:
@wgliang PTAL, thanks!

hoegaarden commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-e2e-gce

hoegaarden commented:
@wgliang Would you be able to review again? Thanks!

wgliang commented on pkg/cloudprovider/providers/vsphere/vclib/connection_test.go:
```diff
@@ -0,0 +1,239 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
2016->2018

wgliang commented on pkg/cloudprovider/providers/vsphere/vclib/connection_test.go:
```diff
@@ -0,0 +1,239 @@
+	}
+
+	// calculate the leaf certificate's fingerprint
+	x509LeafCert := server.TLS.Certificates[0].Certificate[0]
```
You should judge the validity of the index and avoid panic.

hoegaarden replied:
We set the certificate ourselves [just above](https://github.com/kubernetes/kubernetes/pull/64758/files/de5711a8cc66b4f52a94f3acb8fb8249d03317bf#diff-d4d8a303b372aefe62300cab90ebba2dR62), so I guess we didn't really think about that.
But we can definitely make that a safer.

wgliang commented on pkg/cloudprovider/providers/vsphere/vclib/connection_test.go:
```diff
@@ -0,0 +1,239 @@
+limitations under the License.
+*/
+
+package vclib_test
```
why the name is vclib_test, but not vclib?

wgliang commented on pkg/cloudprovider/providers/vsphere/vclib/fixtures/fixtures.go:
```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
nit

hoegaarden committed:
Update copyright header

hoegaarden committed:
Check if the server actually configured with a certificate

hoegaarden commented:
@wgliang PTAL

hoegaarden commented:
@kubernetes/cloud-provider-vsphere-maintainers @wgliang @dougm 
Does anyone want to review and give us feedback on that?

wgliang commented:
/cc @divyenpatel

dougm commented:
Overall lgtm, just wondering if we can avoid `normalizeThumbprint` or if there is a valid end-user use-case maybe it should be in govmomi instead.  And using the simulator may simplify some of the tests, but just a suggestion to look at, not a requirement.

hoegaarden commented:
Thanks for the feedback, @dougm !
- We can avoid `normalizeThumbprint`, it is not strictly necessary but sort of a UX improvement for users. Our tests started out generating lowercase stringified thumbprints, so we ran into this UX "issue" ourselves right away. We can do (a combination) of the following -- please let us know which (combination) you prefer
  - remove `normalizeThumbprint` from this PR
  - add documentation for the format of the thumbprint
  - add syntactical validation when reading the configuration file in the cloudprovider code
  - add syntactical validation when configuring a thumbprint in govmomi
  - use some normalisation in govmomi (and remove from the cloudprovider code)
- We learned about the simulator only a bit later, so the initial tests were already setup with a plain https server. If you have good reasons why we should not do this, let us know, otherwise we keep it like this.

dougm commented:
I do think we should remove `normalizeThumbprint`.  User can get the thumbprint from their web browser or one of the commands below for example.  They all use the same format, as do the related VMware API fields.

```
govc about.cert -k -u $target -thumbprint
```

```
openssl s_client -connect $target </dev/null 2>/dev/null | openssl x509 -fingerprint -noout
```

dougm commented:
I think the tests are fine as-is, just wanted to mention simulator as it might be a better option for future tests.

hoegaarden committed:
Do not do noramlization of the fingerprint format

hoegaarden committed:
Remove unused exported errors

hoegaarden committed:
Update bazel

... after rebase

hoegaarden commented:
/test pull-kubernetes-e2e-gce

hoegaarden commented:
PTAL

dougm commented:
Thanks @mariantalla @hoegaarden lgtm

/approve

hoegaarden commented:
@dougm @divyenpatel @wgliang Is there anything left to do for us to get the *lgtm* -- i think that's missing to get this merged, isn't it?

divyenpatel commented:
/assign @divyenpatel

divyenpatel commented:
/lgtm

dougm commented:
/test pull-kubernetes-kubemark-e2e-gce-big

hoegaarden commented:
/test pull-kubernetes-e2e-gce

mariantalla commented:
/test pull-kubernetes-e2e-gce

hoegaarden commented:
/test pull-kubernetes-e2e-gce

hoegaarden commented:
/test pull-kubernetes-e2e-gce

hoegaarden commented:
/test pull-kubernetes-e2e-gce

marpaia commented:
/sig vmware
